src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,180 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
+                                "Principal"))};
+      absl::c_copy(declaration.arguments(), std::back_inserter(arguments));
+      can_say_declarations.push_back(datalog::RelationDeclaration(
+          absl::StrCat("canSay_", declaration.relation_name()), false,
+          arguments));
+      declaration = datalog::RelationDeclaration(
+          absl::StrCat("canSay_", declaration.relation_name()), false,
+          arguments);
+    }
+  }
+  return can_say_declarations;
+}
+
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::SaysExtendRelationDeclarations(
+    const std::vector<datalog::RelationDeclaration>& transformed_declarations) {
+  // The translated declarations are all extended with "says_" and a speaker
+  // argument
+  std::vector<datalog::RelationDeclaration>
+      says_extended_relational_declarations;
+  for (const datalog::RelationDeclaration& declaration :
+       transformed_declarations) {
+    std::vector<datalog::Argument> arguments = {datalog::Argument(
+        "speaker", datalog::ArgumentType(
+                       datalog::ArgumentType::Kind::kPrincipal, "Principal"))};
+    absl::c_copy(declaration.arguments(), std::back_inserter(arguments));
+    says_extended_relational_declarations.push_back(
+        datalog::RelationDeclaration(
+            absl::StrCat("says_", declaration.relation_name()), false,
+            arguments));
+  }
+  return says_extended_relational_declarations;
+}
+
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::RelationDeclarationToDLIR(
+    const std::vector<datalog::RelationDeclaration>& relation_declarations) {
+  // Note that the additional relation declarations that are generated for
+  // (possibly nested) cansay expressions are based on an environment
+  // that tracks the maximum depth with which they appear. This environment is
+  // populated by side-effect while translating the program body. As a result,
+  // it is assumed that this function is called after the body is translated.
+
+  std::vector<datalog::RelationDeclaration> transformed_declarations =
+      LoweringToDatalogPass::TransformDeclarations(relation_declarations);
+
+  // Produce a mapping from predicate names to predicate typings
+  // (where a predicate typing is the same as a predicate relation declaration)
+  absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>
+      type_environment;
+  for (const datalog::RelationDeclaration& declaration :
+       transformed_declarations) {
+    type_environment.insert({declaration.relation_name(), declaration});
+  }
+
+  std::vector<datalog::RelationDeclaration> can_say_declarations =
+      LoweringToDatalogPass::CanSayDeclarations(type_environment);
+  absl::c_copy(can_say_declarations,
+               std::back_inserter(transformed_declarations));
+
+  // The transformed relation declarations are the same as the ones from the
+  // surface program plus another one for "canActAs"
```

Comment:
Add a comment about what datalog relation or datalog declaration is added.  Is it the following?

```C++
.decl  canActAs(P1: Principal, p2: Principal)
```